// @thirdweb-dev/wallets/evm/connectors/token-bound-smart-wallet/dist/thirdweb-dev-wallets-evm-connectors-token-bound-smart-wallet.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ethers = require('ethers');
var evm_connectors_smartWallet_dist_thirdwebDevWalletsEvmConnectorsSmartWallet = require('../../smart-wallet/dist/thirdweb-dev-wallets-evm-connectors-smart-wallet.cjs.prod.js');
var constants = require('../../../../dist/constants-50e5d28e.cjs.prod.js');
require('../../../../dist/connector-1b2fa06d.cjs.prod.js');
require('eventemitter3');
require('../../../../dist/url-8b7f19db.cjs.prod.js');
require('@thirdweb-dev/chains');
require('../../../../dist/headers-b16c2eb6.cjs.prod.js');
require('@account-abstraction/contracts');
require('../../../../dist/defineProperty-b749763b.cjs.prod.js');
require('@thirdweb-dev/sdk');
require('../../../wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.cjs.prod.js');

class TokenBoundSmartWalletConnector extends evm_connectors_smartWallet_dist_thirdwebDevWalletsEvmConnectorsSmartWallet.SmartWalletConnector {
  constructor(input) {
    super({
      ...input,
      factoryAddress: input.registryAddress || constants.ERC6551_REGISTRY
    });
    this.tbaConfig = input;
    // TODO default account implementation address
  }
  defaultFactoryInfo() {
    return {
      createAccount: async factory => {
        return factory.prepare("createAccount", [this.tbaConfig.accountImplementation, this.chainId, this.tbaConfig.tokenContract, this.tbaConfig.tokenId, this.tbaConfig.salt, ethers.ethers.utils.toUtf8Bytes("")]);
      },
      getAccountAddress: async factory => {
        return await factory.call("account", [this.tbaConfig.accountImplementation, this.chainId, this.tbaConfig.tokenContract, this.tbaConfig.tokenId, this.tbaConfig.salt]);
      }
    };
  }
}

exports.TokenBoundSmartWalletConnector = TokenBoundSmartWalletConnector;
